### PR TITLE
[HIP] [CI] [JIT] [ROCm][MoE] Add fused MoE routing preamble for MXFP4 decode

### PR DESCRIPTION
**Pull request by sshlyapn**
## Motivation

At decode shapes the MXFP4 MoE routing preamble is four short kernels back to back, and the gaps between them cost more than the work inside them. Each kernel is only 5-7us, and every boundary is a full round trip through HBM plus a launch.

This fuses the whole sequence into one kernel (or two kernels above ~104 tokens). End to end this brings up to 9% gain at low concurrency, where the preamble is a meaningful share of each decode step.

The target model is [upstage/Solar-Open2-250B](https://huggingface.co/upstage/Solar-Open2-250B), and the current constraints follow directly from it: gfx950, bf16 in, hidden size 4096, num_experts <= 512, single expert group, biased grouped top-k. Nothing in the design is specific to those values - other hidden sizes, expert counts, topk values, and plain softmax expert selection can be added on top of the same kernel. They are left out here to keep the first version simpler.

<details>
<summary>Performance measurement: M sweep, E=320, topk=8, unit_size=16</summary>

| M | fused (us) | stock (us) | uplift |
|---|---|---|---|
| 1 | 9.19 | 18.28 | 1.99x |
| 32 | 11.60 | 22.42 | 1.93x |
| 64 | 12.83 | 22.65 | 1.76x |
| 103 | 14.63 | 22.96 | 1.57x |
| 104 | 13.59 | 23.07 | 1.70x |
| 128 | 14.08 | 23.31 | 1.66x |

torch.profiler `device_time_sum`, i.e. summed GPU kernel busy time. 

Reproduce with the third command in the test plan below. That table was collected with more iterations than the committed harness uses (it hardcodes `num_iters=10, num_warmup=2`).

</details>

Corresponding PR to vLLM: https://github.com/vllm-project/vllm/pull/56420

## Technical Details

At M=64, E=320, topk=8 the profiler shows:

```
--- STOCK (M=64) ---
  6.72us  aiter::grouped_topk_kernel<hip_bfloat16, float4, 1, true, true, false>
  4.96us  aiter::opus_moe_sorting_entry<MoeSortingMultiPhaseKernel_P0_v2<...>>
  4.60us  aiter::opus_moe_sorting_entry<MoeSortingMultiPhaseKernel_P23<...>>
  5.96us  aiter::fused_mx_quant_moe_sort_kernel<bf16, opus::fp4_t, 256, 16>

--- FUSED (M=64) ---
 13.60us  aiter::fmr::fused_moe_routing_kernel<256, 16, bf16, 0, bf16>
```

So `biased_grouped_topk` -> `moe_sorting` (two phases) -> `fused_dynamic_mxfp4_quant_moe_sort` becomes one launch. Note `moe_sorting` is a single Python call but dispatches two kernels. Optional `moe_buf` zero-fill is folded into the same kernel instead of being a separate memset.

The fused kernel uses a self-resetting grid barrier in place of the kernel boundaries. Above ~104 tokens the grid no longer fits co-resident, so the barrier is replaced by a second launch; both paths share the same device code and differ only in where the barrier sits. Selection is deterministic and the sort is stable, so output is bit-identical to the stock chain for tie-free gating.

The kernel writes the full sorted map, including the region past `num_valid`: the downstream stage-1 GEMM reads an id per block before it consults `num_valid_ids`, so that tail is live memory under a caching allocator.

This is inert on its own - a matching vLLM PR is needed to call it, and will be opened shortly. Callers must check `fused_moe_router_supported()` first and own the workspace via `get_fused_moe_router_workspace()`, since a captured CUDA graph bakes the pointer in and the buffer has to outlive the capture. Expert parallelism is supported through an optional `expert_mask`.

## Test Plan

```bash
# Correctness
python -m pytest op_tests/test_fused_moe_router.py -q

# Stress (multi-GPU, graph capture, concurrent streams)
python op_tests/multigpu_tests/test_fused_moe_router_stress.py

# Benchmark + correctness sweep
python op_tests/test_fused_moe_router.py -m 1,32,64,103,104,128 -ek 320,8 -u 16
```

The stress suite covers what unit tests do not reach: barrier churn across grid shapes, fused/split interleaving, scratch staleness, allocator poisoning, determinism soak, CUDA graph capture and replay, graph/eager interleaving, concurrent streams, CU pressure from a competing GEMM, and multi-device use in one process.

## Test Result

All green on gfx950.

```
327 passed in 77.19s
PASSED: 0 failure(s) over 1 round(s), 12 suite(s)
All fused_moe_router tests passed!   (RC=0, all error columns zero)
```

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: aiter/fused_moe.py [hunk 1]
```diff
@@ -1080,6 +1107,335 @@ def _fused_moe_impl(
...
+    # router only implements the 2-stage non-FLAT path. Derived exactly as
+    # `fused_moe_router` does, so `get_2stage_cfgs` is an lru_cache hit there.
+    activation = ActivationType(activation)
+    quant_type = QuantType(quant_type)
+    gate_mode = GateMode(gate_mode)
+    M = hidden_states.shape[0]
+    E, model_dim, inter_dim = get_inter_dim(w1.shape, w2.shape)
+    dtype, quant_type, q_dtype_a, q_dtype_w = _resolve_quant_dtypes(
+        M, hidden_states, w1, quant_type, activation, gate_mode, a1_scale, dtype
+    )
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply aiter/fused_moe.py:1344]: Fixed

File: csrc/py_itfs_cu/fused_moe_router_entry.cu [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                " B exceeds the per-block limit of ", prop.sharedMemPerBlock,
+                " B (num_tokens=", M, ", topk=", topk, ", num_experts=", E,
+                "); reduce the token count");
+
+    auto* ws_base      = reinterpret_cast<uint8_t*>(workspace.data_ptr());
+    auto* ws_sem       = reinterpret_cast<unsigned int*>(ws_base);
+    auto* ws_tok_scale = ws_base + kTokScaleOffset;
+
+    const opus::bf16_t* g = reinterpret_cast<const opus::bf16_t*>(gating.data_ptr());
+    const opus::bf16_t* h = reinterpret_cast<const opus::bf16_t*>(hidden.data_ptr());
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply csrc/py_itfs_cu/fused_moe_router_entry.cu:254]: Added

File: aiter/ops/fused_moe_router.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    Constraints checked by the C++ entry: ``hidden.shape[1] == 4096``,
+    ``num_experts <= 512``, ``unit_size`` a power of two, bf16 inputs,
+    ``num_expert_group == topk_group == 1``.
+
+    Args:
+        gating: ``[M, num_experts]`` bf16 router logits.
+        bias: ``[num_experts]`` bf16 e_score_correction_bias.
+        hidden: ``[M, cols]`` bf16 activations to quantize.
+        topk_ids: ``[M, topk]`` int32 output.
```
[the remark this answers is not in the document]
Comment (sshlyapn) [reply aiter/ops/fused_moe_router.py]: Added

**Comment by sshlyapn** [comment]
@yzhou103 could you please take a look at this PR?

**Comment by yzhou103** [comment]
please resolve conflicts

**Comment by heslami** [comment]
On the conflicts: we hit the same ones building on `dd2fabb51` and have resolved them, so
the resolution is available if it saves you the work. Main is 120 commits ahead, but the
collision is small, three hunks in two files:

* `.github/scripts/split_tests.sh`: main's timing table is a superset. Take it wholesale and
  re-insert `test_fused_moe_router.py=95` in sorted position.
* `aiter/fused_moe.py`, signature: main's `_stage2_override`/`routing_num_experts` and your
  `a1_prequant`/`_metadata` are independent additions, so both are kept.
* `aiter/fused_moe.py`, dispatch body: main's activation-parameter normalization and its four
  new `get_2stage_cfgs` arguments, wrapped in your `_metadata` short-circuit, with the
  `a1_prequant` branch restored ahead of the prequant test. One judgement call worth your
  eye: we let main's Opus-path prequant forcing apply to caller-supplied `_metadata` too,
  since it repairs a combination GEMM1 cannot consume, and honouring an invalid one would
  fault rather than respect the caller. It is your contract, so say if you would rather it
  stayed inside the computed-metadata branch.

It is the merge commit `ea0b6768d5` on our branch, and we are happy to open it as a PR against
your branch instead if that is easier. Not built on hardware yet.

Separately, we reproduced this on a different shape (256 routed + 1 shared, top-k 8, hidden
6144, MXFP4, gfx950) and it holds up. Thanks for writing it. The caller picks up the new
`supported()` signature automatically because it filters kwargs by signature, so
`global_num_experts` needed no change on our side.

We have opened https://github.com/ROCm/aiter/pull/5517 to extend this kernel and https://github.com/sgl-project/sglang/pull/39529 to add the SGLang
caller, since nothing in SGLang calls the fused entry today. Both build on and credit this
PR. Happy to fold them in here instead if you would rather keep it in one place.

Three findings bear on the numbers in this PR:

**1. The kernel's cost tracks expert diversity, and the generated gating does not vary it.**

* Padded work is set by how many distinct expert units a batch touches, not by token count,
  since each touched unit pads to 32 rows. Token count is a proxy, and only as good as the
  input's diversity.
* The generated gating picks close to the same experts each time, so a given slot count
  lands on few units. On our shape it touched ~19 units where routing captured from a running
  server touched ~71.5: 2,288 padded rows against 608 at the same token count, 3.8x the work.
* That changes the conclusion rather than scaling it. On generated routing the fused path
  looks uniformly faster; on captured routing it is slower than the unfused chain above ~8
  tokens, and no correction factor recovers the right answer.
* We cannot tell from here whether the M=1..256 table is affected, since that depends on the
  diversity of the model you tuned against. The mechanism applies to any model whose routing
  varies per token; the size of the effect is shape-specific.
* We changed our benchmark to take routing from captured serving tensors, and to check the
  work before reporting the time. It will not emit a timing unless the padded sorted row
  count it actually ran is corroborated four ways. **Three of the four are readings of one
  counter compiled into the kernel** (the count the captured routing produced when it was
  recorded in serving, the mean over the served distribution, and the count the timed run
  reports), so their agreement is a consistency check rather than corroboration. **The
  fourth comes from the incumbent chain's own sorting kernels, shares no source with the
  fused kernel, and is the only independent one.** It is asserted wherever both paths select
  the same experts; at a width where they select different experts, which is an exact score
  tie two selection algorithms resolve differently, the counts are expected to differ and
  the row records that instead of asserting. Across our table: seven asserted and passing,
  one not applicable, none refused on it.
* It refuses rather than estimating, which is worth having: it caught several defects in our
  own capture path that had left the benchmark running and reporting normally. Per-row cost
  then matches in-server cost to 2.3% with no fitted factors, measured against a shortened
  copy of the model and not yet re-established on the full network.

**2. The launch grid is sized for routing spread, and ours is much wider than yours.**

* `max(M, 16)` is the same proxy as above, and it holds while expert diversity stays low,
  since padded rows then grow in step with tokens. It is the right choice for the input it
  was tuned against.
* On our shape rows grow far faster than tokens: at M=16 the kernel walks 2,816 padded rows,
  so 16 workgroups serialize ~176 each.
* We replaced the proxy with the quantity it stands for,
  `clamp(ceil(estimated_rows / 20), 16, CUs / 2)`, computable host-side from values the entry
  already has. That takes the kernel body from 28.83 to 13.02 µs at M=16, entirely in the
  serialized phase, and restores an advantage across our measured range: ~2x up to M=17,
  narrowing to 1.12x by M=59 as spread grows.
* **On a concentrated distribution the two policies largely agree**, which is what we would
  expect, since that is where the proxy is accurate: the lower clamp holds at small M and the
  two track each other across the range, coinciding at several points. **We have not run it
  on your configuration and claim no improvement there.** If you can share the routing you
  tuned against, we will run the same sweep and report whatever it says.

**3. A tie-break rule cannot match the chain's tie order.**

* At bit-exact score ties the fused kernel and the unfused chain select different but
  equally valid experts with identical weights.
* We tried matching the chain's order and withdrew it: the chain keeps the lowest expert ids
  when a subset ties and the highest when all tie, in an order that shifts with expert count.
* That order comes from its vectorized scan rather than a rule, so reproducing it would mean
  reproducing the scan's lane layout.
* We would suggest documenting the op's contract as final-output equivalence, which does hold
  byte-for-byte in our testing, rather than intermediate byte-equality.

On enablement: we measured end-to-end and gated the caller to the token counts where serving
shows a win, rather than enabling unconditionally. The advantage narrows as routing spread
grows with token count, and we measured a small regression at the top of our range. The
per-concurrency table is in https://github.com/sgl-project/sglang/pull/39529.

File: csrc/py_itfs_cu/fused_moe_router_entry.cu [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    const HipDeviceGuard device_guard(gating.device().index());
+    const hipStream_t stream = at::hip::getCurrentHIPStream();
+
+    // Grid width multiplies barrier cost, so size it to the work and keep it
+    // under num_cu so every block stays resident. Phase 1 wants one block per
+    // token; phase 3's rows do not shrink with M, so small M still needs width
+    // -- hence the floor.
+    int GRID   = std::max(M, 16);
+    int dev_id = 0;
+    HIP_CALL(hipGetDevice(&dev_id));
```
Comment (yzhou103) [inline csrc/py_itfs_cu/fused_moe_router_entry.cu:190]: use get_num_cu_func()

**Comment by yzhou103** [comment]
1. please restructure the code:
image
2. Also we're removing torch dependency in our kernel, you can refer to quant_kernels.cu for example.

File: csrc/py_itfs_cu/fused_moe_router_entry.cu (hunk 4 above)
Comment (sshlyapn) [reply csrc/py_itfs_cu/fused_moe_router_entry.cu]: Applied, thanks!

**Comment by sshlyapn** [comment]
@yzhou103 thank you for the comments! I've restructured the code and rebased to the latest main

**Comment by sshlyapn** [comment]
@yzhou103 , could you please take another look at this PR when you have a chance?
